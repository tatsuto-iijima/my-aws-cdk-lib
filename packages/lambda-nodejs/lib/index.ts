import * as cdk from '@aws-cdk/core';
import * as lambdaNodejs from '@aws-cdk/aws-lambda-nodejs';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as fs from 'fs';
const jsSHA = require("jssha");

export interface LambdaNodejsProps {
  func?: lambdaNodejs.NodejsFunctionProps,
  role?: {
    description?: string;
    managedPolicies?: string[];
  },
  version?: {
    removalPolicy: cdk.RemovalPolicy;
  };
}

export class LambdaNodejs extends cdk.Construct {
  public readonly role: iam.Role;

  public readonly nodejsFunction: lambdaNodejs.NodejsFunction;

  public readonly version: lambda.Version;

  constructor(scope: cdk.Construct, id: string, props: LambdaNodejsProps = {}) {
    super(scope, id);

    const finalLambdaNodejsProps: any = {};
    Object.assign(finalLambdaNodejsProps, props.func);
    
    if (props.role) {
      this.role = new iam.Role(this, 'Role', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        description: props.role.description,
      });
      if (props.role.managedPolicies) {
        for (const mangagedPolicie of props.role.managedPolicies) {
          this.role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName(mangagedPolicie));
        }
      }
      finalLambdaNodejsProps.role = this.role;
    }

    this.nodejsFunction = new lambdaNodejs.NodejsFunction(this, 'Function', finalLambdaNodejsProps);

    if (props.version) {
      let entry = './lib/index.Function.ts'
      if (props.func?.entry) entry = props.func.entry;
      const shaObj = new jsSHA("SHA-256", "TEXT", { encoding: "UTF8" });
      shaObj.update(fs.readFileSync(entry).toString());
      const hash = shaObj.getHash("HEX");
      this.version = new lambda.Version(this, 'Version-' + hash, {
        lambda: this.nodejsFunction,
        removalPolicy: props.version.removalPolicy,
      });
    }

  }

}
