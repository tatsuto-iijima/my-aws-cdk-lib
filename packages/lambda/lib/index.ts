import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';

export interface LambdaProps {
  func: lambda.FunctionProps;
  role?: {
    servicePrincipalNames: string[];
    description?: string;
    managedPolicieNames?: string[];
  };
  version?: {
    name: string;
    removalPolicy?: cdk.RemovalPolicy;
  };
}

export class Lambda extends cdk.Construct {
  public readonly role: iam.Role;

  public readonly func: lambda.Function;

  public readonly version: lambda.Version;

  constructor(scope: cdk.Construct, id: string, props: LambdaProps) {
    super(scope, id);

    const finalLambdaProps: any = {};
    Object.assign(finalLambdaProps, props.func);
    
    if (!finalLambdaProps.role && props.role) {
      const finalRoleProps:any = {};
      const servicePrincipals: iam.ServicePrincipal[] = [];
      for (const servicePricipalName of props.role.servicePrincipalNames) {
        servicePrincipals.push(new iam.ServicePrincipal(servicePricipalName));
      }
      finalRoleProps.assumedBy = new iam.CompositePrincipal(...servicePrincipals);
      if (props.role.description) finalRoleProps.description = props.role.description;
      if (props.role.managedPolicieNames) {
        const managedPolicies: iam.IManagedPolicy[] = [];
        for (const mangagedPolicieName of props.role.managedPolicieNames) {
          managedPolicies.push(iam.ManagedPolicy.fromAwsManagedPolicyName(mangagedPolicieName));
        }
        finalRoleProps.managedPolicies = managedPolicies;
      }
      this.role = new iam.Role(this, 'Role', finalRoleProps);
      finalLambdaProps.role = this.role;
    }

    this.func = new lambda.Function(this, 'Function', finalLambdaProps);

    if (props.version) {
      const finalVersionProps:any = {
        lambda: this.func,
      };
      if (props.version.removalPolicy) finalVersionProps.removalPolicy = props.version.removalPolicy;
      this.version = new lambda.Version(this, 'Version-' + props.version.name, finalVersionProps);
    }

  }

}
