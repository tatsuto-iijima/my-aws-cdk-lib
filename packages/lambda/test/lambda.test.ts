import { expect as expectCDK, countResources, ABSENT, arrayWith, objectLike, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
const jsSHA = require('jssha');
import * as fs from 'fs';
import * as AwsCdkLibLambda from '../lib/index';

const shaObj = new jsSHA("SHA-256", "TEXT", { encoding: "UTF8" });
shaObj.update(fs.readFileSync('./test/sample/code.zip').toString());
const codeSha256 = shaObj.getHash("HEX");

describe('Test Construct "Lambda" by default', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambda.Lambda(stack, 'MyTestConstruct', {
    func: {
      code: lambda.Code.fromAsset('./test/sample/code.zip'),
      handler: 'index',
      runtime: lambda.Runtime.NODEJS_12_X,
    },
  });

  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });

  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });

  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          }),
        ),
      },
    }));
  });

  test('IAM Role Property "ManagedPolicyArns" has "AWSLambdaBasicExecutionRole"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          ],
        ]
      })),
    }));
  });

  test('Lambda Version x0 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 0));
  });

  test('Snapshot', () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

});

const testRolePropsData = [
  { servicePrincipalNames: [ 'edgelambda.amazonaws.com' ] },
  { servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ] },
  {
    servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
    description: 'Test Construct "Lambda"',
  },
  {
    servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
    managedPolicieNames: [ 'service-role/AWSLambdaBasicExecutionRole' ],
  },
  {
    servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
    managedPolicieNames: [ 'service-role/AWSLambdaBasicExecutionRole', 'AmazonS3ReadOnlyAccess' ],
  },
  {
    servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
    description: 'Test Construct "Lambda"' ,
    managedPolicieNames: [ 'service-role/AWSLambdaBasicExecutionRole', 'AmazonS3ReadOnlyAccess' ],
  },
];
const testRolePropsTable = [];
for (const testRoleProps of testRolePropsData) {
  testRolePropsTable.push([
    JSON.stringify(testRoleProps),
    testRoleProps,
  ]);
}
describe.each(testRolePropsTable)('Test Construct "Lambda" by option { role: %j }', (json, testRoleProps) => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  if (typeof testRoleProps != 'string') {
    new AwsCdkLibLambda.Lambda(stack, 'MyTestConstruct', {
      func: {
        code: lambda.Code.fromAsset('./test/sample/code.zip'),
        handler: 'index',
        runtime: lambda.Runtime.NODEJS_12_X,
      },
      role: testRoleProps,
    });
  } else {
    throw new Error('ERROR: Type of RoleProps is string');
  }
  
  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });
  
  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });

  test('IAM Role Property "AssumeRolePolicyDocument" has "' + testRoleProps.servicePrincipalNames.join(', ') + '"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: testRoleProps.servicePrincipalNames.length > 1 ? testRoleProps.servicePrincipalNames : testRoleProps.servicePrincipalNames[0],
            },
          }),
        ),
      },
    }));
  });

  if (testRoleProps.description) {
    test('IAM Role Property "Description" is "' + testRoleProps.description + '"', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        Description: testRoleProps.description,
      }));
    });
  } else {
    test('IAM Role Property "Description" is none', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        Description: ABSENT,
      }));
    });
  }

  if (testRoleProps.managedPolicieNames) {
    test.each(testRoleProps.managedPolicieNames)('IAM Role Property "ManagedPolicyArns" is "%s"', (managedPoliciesName) => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        ManagedPolicyArns: arrayWith(objectLike({
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/" + managedPoliciesName,
            ],
          ]
        })),
      }));
    });
  } else {
    test('IAM Role Property "ManagedPolicyArns" is none', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        ManagedPolicyArns: ABSENT,
      }));
    });
  }

});

describe.each([
  [ { name: codeSha256 } ],
  [ { name: codeSha256, removalPolicy: cdk.RemovalPolicy.DESTROY } ],
  [ { name: codeSha256, removalPolicy: cdk.RemovalPolicy.RETAIN } ],
])('Test Construct "Lambda" by option { version: %j }', (option) => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambda.Lambda(stack, 'MyTestConstruct', {
    func: {
      code: lambda.Code.fromAsset('./test/sample/code.zip'),
      handler: 'index',
      runtime: lambda.Runtime.NODEJS_12_X,
    },
    version: option,
  });

  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });

  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });

  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          }),
        ),
      },
    }));
  });

  test('IAM Role Property "ManagedPolicyArns" has "AWSLambdaBasicExecutionRole"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          ],
        ]
      })),
    }));
  });

  test('Lambda Version x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 1));
  });
  
});

describe('Test Construct "Lambda" by option various', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambda.Lambda(stack, 'MyTestConstruct', {
    func: {
      code: lambda.Code.fromAsset('./test/sample/code.zip'),
      handler: 'index',
      runtime: lambda.Runtime.NODEJS_12_X,
    },
    role: {
      servicePrincipalNames: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
      description: 'Test Construct "Lambda"' ,
      managedPolicieNames: [ 'service-role/AWSLambdaBasicExecutionRole', 'AmazonS3ReadOnlyAccess' ],
    },
    version: {
      name: codeSha256,
    },
  });

  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });

  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });

  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: [ 'lambda.amazonaws.com', 'edgelambda.amazonaws.com' ],
            },
          }),
        ),
      },
    }));
  });

  test('IAM Role Property "Description" is "Test Construct "Lambda""', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      Description: 'Test Construct "Lambda"',
    }));
  });

  test('IAM Role Property "ManagedPolicyArns" has "AWSLambdaBasicExecutionRole"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          ],
        ]
      })),
    }));
  });

  test('IAM Role Property "ManagedPolicyArns" has "AmazonS3ReadOnlyAccess"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/AmazonS3ReadOnlyAccess",
          ],
        ]
      })),
    }));
  });

  test('Lambda Version x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 1));
  });

  test('Snapshot', () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

});
