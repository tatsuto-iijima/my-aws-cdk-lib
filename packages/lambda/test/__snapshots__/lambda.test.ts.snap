// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Construct "Lambda" by default Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bArtifactHash98CCFFF6": Object {
      "Description": "Artifact hash for asset \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3Bucket03FE7177": Object {
      "Description": "S3 bucket for asset \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB": Object {
      "Description": "S3 key for asset version \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructFunction4795DE86": Object {
      "DependsOn": Array [
        "MyTestConstructFunctionServiceRole786CED66",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3Bucket03FE7177",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructFunctionServiceRole786CED66",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructFunctionServiceRole786CED66": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Test Construct "Lambda" by option various Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bArtifactHash98CCFFF6": Object {
      "Description": "Artifact hash for asset \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3Bucket03FE7177": Object {
      "Description": "S3 bucket for asset \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
    "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB": Object {
      "Description": "S3 key for asset version \\"8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843b\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructFunction4795DE86": Object {
      "DependsOn": Array [
        "MyTestConstructRole16E7E65D",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3Bucket03FE7177",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8c54104f8c241263dca6257bd81975a52f19a118bb01b5596f292e4bd46d843bS3VersionKey8E7920FB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructRole16E7E65D",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructRole16E7E65D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                  "edgelambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Test Construct \\"Lambda\\"",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonS3ReadOnlyAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructVersion04a7b4623c7bbdfb27115c29dbd8c6cab43207b1ce6b50e22e03296c0adaa76469BD4225": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructFunction4795DE86",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
  },
}
`;
